### PR TITLE
Home work2

### PR DESCRIPTION
**Pull request by Allkman**

### CODE REVIEW

File: Matrix.ConsoleApp/Matrix.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+/// [(i==j), (i!=j), (i!=j)] (X, 0, 0) is to index this 'X'
+/// [(i!=j), (i==j), (i!=j)] (0, X, 0)
+/// [(i!=j), (i!=j), (i==j)] (0, 0, X)
+/// </summary>
+namespace Matrix.ConsoleApp
+{
+    internal class Matrix
+    {
+        //1.
+        public int[] DiagonalMatrixArray;
```
Comment (radislavkireev) [inline Matrix.ConsoleApp/Matrix.cs]: make this field as a private. 

If diagonal may be updated at any moment of time, Size is out of sync with it

File: Matrix.ConsoleApp/Matrix.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            //validating if params element is null
+            //https://stackoverflow.com/questions/6584131/c-sharp-is-it-possible-to-have-null-params @Joshua Rodgers answer
+            if (diagonalMatrixArray == null)
+            {
+                _size = 0;
+            }
+            else
+            {
+                //a loop for creating a matrix, iterating through rows and columns:
+                for (int row = 0; row < _size; row++)
```
Comment (radislavkireev) [inline Matrix.ConsoleApp/Matrix.cs]: If you want to display matrix here, use overridden ToString method

File: Matrix.ConsoleApp/Matrix.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    {
+                        Console.Write($"{this[row, column]} ");
+                    }
+                }
+            }
+        }
+        //5.
+        public int Track()
+        {
+            int sum = DiagonalMatrixArray.Sum();
```
Comment (radislavkireev) [inline Matrix.ConsoleApp/Matrix.cs]: simplify 
`return DiagonalMatrixArray.Sum();`

File: Matrix.ConsoleApp/MatrixExtension.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Matrix.ConsoleApp
+{
+    internal static class MatrixExtension 
+    {
+        public static Matrix GetNewMatrixFromAddingTwoMatrices(this Matrix first, Matrix second)
+        {
+            //adding elements of two arrays
+            int[] array = new int[first.Size];
+            if (first.Size > second.Size)
```
Comment (radislavkireev) [inline Matrix.ConsoleApp/MatrixExtension.cs:12]: Complete the requirement: If the dimensions of the matrix do not match, the smaller matrix is padded with zeros.

File: Matrix.ConsoleApp/Matrix.cs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                else if (i != j)
+                {
+                    return 0;
+                }
+                else if (i == j)
+                { 
+                    return DiagonalMatrixArray[i];
+                }
+                else return 0;
+            }
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: Check if i < _size not to have OutOfRangeException

File: Matrix.ConsoleApp/Matrix.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            set => DiagonalMatrixArray[i] = value;
+        }
+        //passing a a single-dimentional array of params as in ms.docs
+        //3.
+        //https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/passing-arrays-as-arguments
+        //https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/params
+        //https://www.c-sharpcorner.com/UploadFile/c63ec5/use-params-keyword-in-C-Sharp/
+        public Matrix(params int[] diagonalMatrixArray)
+        {
+            DiagonalMatrixArray = diagonalMatrixArray;
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: Will be exception of diagonalMatrixArray in null, check before, not on line 77

File: Matrix.ConsoleApp/Matrix.cs [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            _size = diagonalMatrixArray.Length;
+
+            //validating if params element is null
+            //https://stackoverflow.com/questions/6584131/c-sharp-is-it-possible-to-have-null-params @Joshua Rodgers answer
+            if (diagonalMatrixArray == null)
+            {
+                _size = 0;
+            }
+            else
+            {
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: Can't understand what are you doing here. Only output the values of matrix? In constructor? Never output anything in a constructor

File: Matrix.ConsoleApp/Matrix.cs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+            }
+            set => DiagonalMatrixArray[i] = value;
+        }
+        //passing a a single-dimentional array of params as in ms.docs
+        //3.
+        //https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/passing-arrays-as-arguments
+        //https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/params
+        //https://www.c-sharpcorner.com/UploadFile/c63ec5/use-params-keyword-in-C-Sharp/
+        public Matrix(params int[] diagonalMatrixArray)
+        {
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: Copy array diagonalMatrixArray to DiagonalMatrixArray , not only copy the reference

File: Matrix.ConsoleApp/Matrix.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+/// example of 3x3 diagonal matrix
+/// [(i==j), (i!=j), (i!=j)] (X, 0, 0) is to index this 'X'
+/// [(i!=j), (i==j), (i!=j)] (0, X, 0)
+/// [(i!=j), (i!=j), (i==j)] (0, 0, X)
+/// </summary>
+namespace Matrix.ConsoleApp
+{
+    internal class Matrix
+    {
+        //1.
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: Has to be private not to have chance to change array elements directly, not using appropriate indexer

File: Matrix.ConsoleApp/Matrix.cs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                    for (int column = 0; column < _size; column++)
+                    {
+                        Console.Write($"{this[row, column]} ");
+                    }
+                }
+            }
+        }
+        //5.
+        public int Track()
+        {
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: just return DiagonalMatrixArray.Sum();

File: Matrix.ConsoleApp/Matrix.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            return string.Join(", ", DiagonalMatrixArray);
+        }
+        public override bool Equals(object obj)
+        {
+            //according to Jeffrey Richter (CLR via C#)
+            var item = obj as Matrix;
+            if (item == null) //if obj args is null return false
+            {
+                return false;
+            }
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: No need to do this, if item is not Matrix, you will return false in line 111 after check for null after casting

File: TMS.ConsoleApp/Entities/Lecture.cs [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TMS.ConsoleApp.Entities
+{
+    internal class Lecture : EntityBase
+    {
+        public string Topic { get; set; } = string.Empty;
+        public PracticalLesson? PracticalLesson { get; set; }
+        public Lecture(string description, string topic, PracticalLesson practicalLesson)
```
Comment (radislavkireev) [inline TMS.ConsoleApp/Entities/Lecture.cs]: Lecture should not include Practice

> A lecture has a topic (an arbitrary string, possibly empty or equal to null),
>  and 
> a practical lesson - a link to a task condition (an arbitrary string, possibly empty or equal to null)

meaning that lecture has a topic and practice has a task

File: TMS.ConsoleApp/Entities/Training.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using TMS.ConsoleApp.Interfaces;
+
+namespace TMS.ConsoleApp.Entities
+{
+    internal class Training : EntityBase, IDeepClonable<Training>
+    {
+        public List<EntityBase> trainingTypes = new List<EntityBase>();
+        private Training _training;
```
Comment (radislavkireev) [inline TMS.ConsoleApp/Entities/Training.cs]: drop this field

File: Matrix.ConsoleApp/MatrixExtension.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Linq;
+
+namespace Matrix.ConsoleApp
+{
+    internal static class MatrixExtension 
+    {
+        public static Matrix GetNewMatrixFromAddingTwoMatrices(this Matrix first, Matrix second)
+        {
+            //adding elements of two arrays
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/MatrixExtension.cs]: Simplify, do not repeat yourself, combine lines from 11 till 27 in one block, not reassign  memory in 22. Determine the needed memory size before line 11 and not do the operation twice

File: Matrix.ConsoleApp/Matrix.cs [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+/// If the dimensions of the matrix do not match, the smaller matrix is padded with zeros.
+/// </summary>
+/// <summary>
+/// example of 3x3 diagonal matrix
+/// [(i==j), (i!=j), (i!=j)] (X, 0, 0) is to index this 'X'
+/// [(i!=j), (i==j), (i!=j)] (0, X, 0)
+/// [(i!=j), (i!=j), (i==j)] (0, 0, X)
+/// </summary>
+namespace Matrix.ConsoleApp
+{
```
Comment (IgorKheidorov) [inline Matrix.ConsoleApp/Matrix.cs]: Name has to be DiagonalMatrix

File: TMS.ConsoleApp/Entities/Training.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        public Training Clone()
+        {
+            var trainingClone = new Training();
+
+            if (_training != null)
+            {
+                trainingClone._training = _training.Clone(); //reference type .Clone() that does the same
+            }
+           
+            return trainingClone;
```
Comment (radislavkireev) [inline TMS.ConsoleApp/Entities/Training.cs:42]: The clone should contain same Description property and list of EntityBase. 
Current implementation clones only private property of the Training type  and do not see it is used actually .

To make clone, do the ctor and create new instance:
I can`t recall if Igor asked to-do the deep clone, but assume that it is not required. 
`
public Training Clone()
{
    return new Training(this.Description, this.trainingType);  
}
`

**Review by radislavkireev — CHANGES_REQUESTED** [review]

File: TMS.ConsoleApp/Entities/Training.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using TMS.ConsoleApp.Interfaces;
+
+namespace TMS.ConsoleApp.Entities
+{
+    internal class Training : EntityBase, IDeepClonable<Training>
+    {
+        public List<EntityBase> trainingTypes = new List<EntityBase>();
```
Comment (IgorKheidorov) [inline TMS.ConsoleApp/Entities/Training.cs]: No sense in this field, clone directly the object of Training (list of lessons)

File: TMS.ConsoleApp/Entities/Training.cs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            trainingTypes.Add(training);
+        }
+
+        public bool IsPractical()
+        {
+            if (trainingTypes.Count == 0)
+            {
+                return false;
+            }
```
Comment (IgorKheidorov) [inline TMS.ConsoleApp/Entities/Training.cs]: Simplify, just return trainingTypes.All(x => x is PracticalLesson)

File: TMS.ConsoleApp/Entities/Lecture.cs [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TMS.ConsoleApp.Entities
+{
+    internal class Lecture : EntityBase
+    {
+        public string Topic { get; set; } = string.Empty;
```
Comment (IgorKheidorov) [inline TMS.ConsoleApp/Entities/Lecture.cs:11]: PracticalLesson field is not necessary, Lecture and TrainingLecture are different separate not linked each to each classes

File: TMS.ConsoleApp/Entities/Training.cs [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                return true;
+            }
+            return false;            
+        }
+        //https://stackoverflow.com/questions/129389/how-do-you-do-a-deep-copy-of-an-object-in-net
+        public Training Clone()
+        {
+            var trainingClone = new Training();
+
```
Comment (IgorKheidorov) [inline TMS.ConsoleApp/Entities/Training.cs]: You need the deep cloning, not copy only references, create a new List<EntityBase> and copy to it lesson from target List

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
One project- one solution- one PR. You have to make two PR here instead of one

File: TMS.ConsoleApp/Entities/EntityBase.cs [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TMS.ConsoleApp.Entities
+{
+    internal class EntityBase
+    {
+        public string? Desctription { get; set; } = string.Empty;
```
Comment (radislavkireev) [inline TMS.ConsoleApp/Entities/EntityBase.cs]: typo: Des**cr**iption

File: TMS.ConsoleApp/Entities/Training.cs (hunk 16 above)
Comment (radislavkireev) [reply TMS.ConsoleApp/Entities/Training.cs:42]: I was mistaken, Igor is asking for the deep cloning here. 

All the items in trainingTypes list should be cloned as well.

File: TMS.ConsoleApp/Entities/Training.cs [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+            Description = description;
+
+            _trainingTypes = trainingTypes;
+        }
+        public Training() { }
+        public void Add(EntityBase training) 
+        {
+            _trainingTypes.Add(training);
+        }
+        public bool IsPractical()
```
Comment (radislavkireev) [inline TMS.ConsoleApp/Entities/Training.cs:23]: return _trainingTypes.All(i => i is Lecture);

File: TMS.ConsoleApp/Entities/EntityBase.cs [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TMS.ConsoleApp.Entities
+{
+    internal class EntityBase
+    {
+        //if all classes have same prop Description,
+        //I want them to inherit form this base class that will
+        public string? Description { get; set; } = string.Empty;
```
Comment (radislavkireev) [inline TMS.ConsoleApp/Entities/EntityBase.cs:13]: Clone method here

File: TMS.ConsoleApp/Entities/Training.cs (hunk 16 above)
Comment (Allkman) [reply TMS.ConsoleApp/Entities/Training.cs:42]: About Cloning: I created "Copy constructor" where I re-initialize the props:
  public Training(string description, List<EntityBase> trainingTypes)
        {
            Description = description;

            _trainingTypes = trainingTypes;
        }
Implement Clone() like so: (with new I am creating a new memory address in the Heap)
         public Training Clone()
        {
            return new Training(this.Description, this._trainingTypes); // returning new obj of type (Training)
        }

File: Matrix.ConsoleApp/MatrixTracker.cs [hunk 24]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Comment (radislavkireev) [inline Matrix.ConsoleApp/MatrixTracker.cs]: Drop this empty class

File: TMS.ConsoleApp/Entities/Training.cs [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            return _trainingTypes.All(x => x is PracticalLesson);
+        }
+        public Training Clone()
+        {
+            return new Training(this.Description, this._trainingTypes);
```
Comment (radislavkireev) [inline TMS.ConsoleApp/Entities/Training.cs]: The clone should be deep cloning, meaning that each element of trainingType should be cloned as well, 
it should not be passed as  a reference.

File: Task1.GenericDiagonalMatrix/SquareDiagonalMatrix.cs [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+                if (i >= 0 && j >= 0 || i <= _size && j <= _size || i == j)
+                {
+                    _diagonalMatrixArray[i] = value;
+                }
+
+            }
+        }
+        public SquareDiagonalMatrix(params T[] squareDiagonalMatrix)
+        {
+
```
Comment (radislavkireev) [inline Task1.GenericDiagonalMatrix/SquareDiagonalMatrix.cs]: Size and _diagonalMatrixArray should be set here

**Review by radislavkireev — CHANGES_REQUESTED** [review]

**Review by radislavkireev — APPROVED** [review]

**Pull request merged**